### PR TITLE
Speed up builds

### LINKED ISSUE #1915: Investigate include-what-you-use to improve build times
### Description of the task
https://include-what-you-use.org/ is very cool and might help with our build times. We can probably run this once and get some benefit and periodically run it again. It would be nice to have a script that we can reuse later with supporting documentation.

### Acceptance criteria

- [ ] Run this tool and improve build times and quantify the change
- [ ] make some script/documentation so it can be run again in the future

### LINKED ISSUE #2144: Re-enable cubemx regen check
### Description of the task
as titled, see TODO in code, we need to make sure it doesn't cause CI to fail as it did here: https://github.com/UBC-Thunderbots/Software/pull/2100/checks?check_run_id=2810567643

### Acceptance criteria

- [ ] Item A
- [ ] Item B
- [ ] Item C

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Split dynamic parameters into header and source
* Changed some includes
* re-enabled cubemx check
* deleted c dynamic params (didn't want to bother splitting into header and source and it's unused)

Results in ~30% reduction in CI time

### Resolved Issues
resolves #2144
resolves #1915

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/parameter/generation_scripts/cpp_header_config.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                param.command_line_option_entry_with_prefix(arg_prefix + ".")
+            )
+            self.included_config_load_command_line_args_into_config_contents.append(
+                param.load_command_line_arg_into_config_with_dependencies(
+                    load_dependency, arg_prefix + "."
+                )
+            )
+
+        # top level config has access to all configs, so no need to recursively add options
+        # if not self.is_top_level_config:
```
Comment (kelvinkoon) [inline src/shared/parameter/generation_scripts/cpp_header_config.py]: I can't tell if this is commented code that should be removed or a part of the above comment.
If it's part of the above comment, remove the `:` at the end or use ``` to block comment.

File: src/shared/parameter/generation_scripts/cpp_source_config.py [hunk 2]
```diff
@@ -173,11 +149,13 @@ def dependency_graph(self, dependency_graph: nx.DiGraph):
...
+    def dfs_helper(
+        self, config: CppSourceConfig, arg_prefix: str, load_dependency: str
+    ):
         """A depth first search helper for adding the necessary prefix to accessing and setting
         parameters of included configs in loadFromCommmandLineArguments function
 
-        :param config: the current CppConfig object
+        :param config: the current CppSourceConfig object
         :param arg_prefix: the prefix for accessing the arg struct
         :param load_depndency: the prefix for accessing the actual parameter
```
Comment (kelvinkoon) [inline src/shared/parameter/generation_scripts/cpp_source_config.py]: Misspelled `load_dependency`. Please kindly fix as well.

File: src/shared/parameter/generation_scripts/cpp_writer.py [hunk 3]
```diff
@@ -47,30 +45,107 @@
...
 class CppWriter(object):
     @staticmethod
-    def create_config_list_from_metadata(
+    def create_header_config_list_from_metadata(
+        top_level_config_name: str, config_metadata: dict
+    ) -> List[CppHeaderConfig]:
+        """Takes the config metadata loaded by config_yaml_loader, and converts it to a list of CppHeaderConfig objects;
+        this includes setting the dependency graphs needed for the configs.
+
+        :param top_leve_config_name: the name of the top level config
```
Comment (kelvinkoon) [inline src/shared/parameter/generation_scripts/cpp_writer.py]: Misspelled `top_level_config_name`.

File: src/software/simulation/threaded_simulator_test.cpp [hunk 4]
```diff
@@ -146,7 +146,7 @@ TEST_F(ThreadedSimulatorTest, add_robots_and_primitives_while_simulation_running
     threaded_simulator.startSimulation();
     // yield and sleep to give the simulation thread the best chance of running
     std::this_thread::yield();
-    std::this_thread::sleep_for(std::chrono::milliseconds(1500));
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
```
Comment (kelvinkoon) [inline src/software/simulation/threaded_simulator_test.cpp:149]: Why the change in values?

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Nice work, just a few nits.

File: src/software/simulation/threaded_simulator_test.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/simulation/threaded_simulator_test.cpp:149]: makes the test pass more often (it's multithreaded)

**Review by akhilveeraghanta — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]
